[#50] 공연장, 공연 홀 정보 수정 기능 병합

tlsgkr7416 opened the pull request:
1. 공연장 수정, 공연 홀 수정, 공연 홀 삭제 기능 병합
2. VenueUpdateRequest DTO로 공연장, 공연 홀 수정 정보 요청
  *프론트에서 공연장이 수정됐는지 안됐는지 boolean 값으로 보내줌(true는 수정한 공연장 데이터, false는 수정 전 공연장 데이터)
  => false는 수정하지 않는 공연장 데이터이므로 DB에 요청을 안함
  *공연장 정보 요청
  *여러 공연 홀을 수정할 수 있어서 수정된 내용을 List로 요청
  *여러 공연 홀을 삭제할 수 있어서 삭제할 공연 홀 id를 List로 요청
 3. 공연장 수정, 공연 홀 수정, 공연 홀 삭제를 하나의 트랜잭션으로 처리
4. 공연 홀 update를 동시에 여러 개 할 수 있도록 DataSource url에 allowMultiQueries=true 설정

tlsgkr7416 committed:
[#50] 공연장, 공연 홀 정보 수정 기능 병합

1. 공연장 수정, 공연 홀 수정, 공연 홀 삭제 기능 병합
2.VenueUpdateRequest DTO로 공연장, 공연 홀 수정 정보 요청
*프론트에서 공연장이 수정됐는지 안됐는지 boolean 값으로 보내줌(true는 수정한 공연장 데이터, false는 수정 전 공연장 데이터)
=> false는 수정하지 않는 공연장 데이터이므로 DB에 요청을 안함
*공연장 정보 요청
*여러 공연 홀을 수정할 수 있어서 수정된 내용을 List로 요청
*여러 공연 홀을 삭제할 수 있어서 삭제할 공연 홀 id를 List로 요청
3. 공연장 수정, 공연 홀 수정, 공연 홀 삭제를 하나의 트랜잭션으로 처리
4. 공연 홀 update를 동시에 여러 개 할 수 있도록 DataSource url에 allowMultiQueries=true 설정

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/VenueController.java:
```diff
@@ -32,8 +33,8 @@ public void insertVenue(@RequestBody @Valid ShowPlace showPlace) {
-    public void updateVenueInfo(@PathVariable int venueId, @RequestBody @Valid Venue venueUpdateRequest) {
-        venueService.updateVenueInfo(venueId, venueUpdateRequest);
+    public void updateVenueInfo(@PathVariable int venueId, @RequestBody @Valid VenueUpdateRequest venueUpdateRequest) {
+        venueService.updateVenueInfo(venueId, venueUpdateRequest, venueUpdateRequest.isVenueInfoChanged());
```
`venueUpdateRequest`를 넘겨주는데 굳이 `isVenueInfoChanged()`도 넘겨줄 필요가 있을까요?

f-lab-dev commented on src/main/java/com/show/showticketingservice/model/venue/VenueUpdateRequest.java:
```diff
@@ -0,0 +1,23 @@
+@AllArgsConstructor
+public class VenueUpdateRequest {
+
+    private final boolean isVenueInfoChanged;
```
`is`는 getter에 붙을거라 필드명에는 안적어주셔도 됩니다~

tlsgkr7416 committed:
[#50] 공연장, 공연 홀 정보 수정 기능 병합

*공연장 DTO 수정
    *기존에 isVenueInfoChanged로 공연장 수정 전 데이터를 보냈을 시 boolean으로 체크를 했지만
    비효율적이어서 수정 전 데이터는 클라이언트에서 데이터 자체를 안 보내는 걸로 수정
    *updateHallIds로 수정할 공연 홀 id를 받아 옴
    *venueHallRequests로 수정할 공연 홀 정보를 받아 옴

tlsgkr7416 committed:
Merge branch 'feature/46/sh' into feature/50/sh

tlsgkr7416 committed:
[#50] 공연장, 공연 홀 수정 기능 테스트 코드 작성

공연장, 공연 홀 서비스 수정 로직을 세부적으로 테스트 기능 작성

tlsgkr7416 committed:
[#49] 공연장, 공연 홀 정보 삭제 기능 병합

1.venueHall테이블에 ON DELETE CASCADE 추가
2.공연장 삭제 기능 테스트 코드 작성

tlsgkr7416 commented on src/test/java/com/show/showticketingservice/service/VenueHallServiceTest.java:
```diff
@@ -0,0 +1,120 @@
+    VenueHallService venueHallService;
+
+    @BeforeEach
+    public void init() {
+        venueHallRequests = new ArrayList<>();
+        venueHallRequestOne = new VenueHallRequest("공연홀A", 50, 60);
+        venueHallRequestTwo = new VenueHallRequest("공연홀B", 30, 50);
+        venueHallRequests.add(venueHallRequestOne);
+        venueHallRequests.add(venueHallRequestTwo);
+
+        updateHallIds = new ArrayList<>();
+        updateHallIds.add(1);
+        updateHallIds.add(2);
+
+        venueId = 1;
+    }
```
update 시키기 전의 데이터입니다.

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/VenueHallService.java:
```diff
@@ -38,24 +37,25 @@ public void checkDuplicationVenueHallName(List<VenueHallRequest> venueHallReques
-        if(!venueHallUpdateRequests.isEmpty()) {
-            checkVenueHallsExists(venueHallUpdateRequests, venueId);
-            venueHallMapper.updateVenueHalls(venueHallUpdateRequests, venueId);
+        if(venueHallRequests != null) {
```
null이 넘어왔으면 예외를 던져줘도 되지 않을까요?

tlsgkr7416 replied:
공연 홀 수정 기능을 변경을 원하지 않을 시 null에서 []로 변경을 했습니다. 그리고 만약 null을 보낸다면 @valid로 체크해줬습니다.

tlsgkr7416 replied:
여기 브랜치의 기능은 공연장 수정, 공연 홀 수정, 공연 홀 삭제가 동시에 일어나서 트랜잭션으로 묶어줬습니다. 만약 클라이언트에서 공연 홀은 수정하길 원하지 않고 공연장 수정이나 공연 홀 삭제를 원할 수 있어서 공연 홀을 수정하길 원하지 않을 시 [](빈 배열)을 보내서 예외 처리는 해주지 않았습니다.  즉 공연 홀 수정만 []이고 공연장 수정 정보나 공연 홀 삭제 정보를 보낸다면 공연 홀은 수정되지 않고 공연장 수정과 공연 홀 삭제 기능만 일어나게 했습니다.

f-lab-dev replied:
변경을 원하지 않을 경우엔 그럼 굳이 `Service`의 코드를 호출할 필요없이 `Controller`에서 처리해줘도 되지 않을까요?

tlsgkr7416 replied:
Controller에서 VenueService를 호출하고 VenueService 내부에서 VenueHallService를 호출합니다. 만약 VenueService의 변경을 원하지 않을 경우 VenueHallService는 변경할 수도 있으므로 service를 호출 시켜줬습니다. 그리고 VenueHallService도 Controller에 따로 호출하려고 했지만 트랜잭션 처리 때문에 VenueService에 호출해서 전체 기능을 하나로 묶어줬습니다.

tlsgkr7416 committed:
[#50]공연 홀 수정과 공연 홀 삭제 시 기능 수정

공연 홀 정보와 삭제할 홀 id는 list 형식으로 받는데 만약 변경을 원하지 않을 경우
null이 아닌 []형식으로 변경.

tlsgkr7416 commented on src/main/java/com/show/showticketingservice/service/VenueHallService.java:
```diff
@@ -46,7 +46,7 @@ public void checkVenueHallsExists(List<VenueHallRequest> venueHallRequests, int
+        if(!venueHallRequests.isEmpty() && !updateHallIds.isEmpty()) {
             checkDuplicationVenueHallName(venueHallRequests);
             checkVenueHallsExists(venueHallRequests, venueId);
             venueHallMapper.updateVenueHalls(updateHallIds, venueHallRequests, venueId);
```
여기 기능에 대해 조금 더 설명 하자면 venueHallRequests는 공연 홀을 수정 할 정보가 담겨 있습니다. 그리고 updateHallIds는 수정 할 공연 홀 id로 db에서 이 홀 id를 찾아서 venueHallRequests로 update 시켜줍니다.

tlsgkr7416 committed:
[#49] 공연장 삭제 기능에 삭제할 id가 존재하지 않는 경우 예외 처리

공연장 삭제 시 id가 존재하지 않을 때 VenueIdNotExistsException 처리

tlsgkr7416 committed:
Merge branch 'feature/49/sh' into feature/50/sh

tlsgkr7416 committed:
[#50] 공연장 공연 홀 병합 기능 수정

* 예외 처리
    1. 공연장 수정 시 존재하지 않는 id로 업데이트할 경우 VenueIdNotExistsException 예외 처리
    2 .공연 홀 수정 시 존재하지 않는 id로 업데이트할 경우 VenueHallIdNotExistsException 예외 처리
    3. 공연 홀 삭제 시 존재하지 않는 id로 삭제할 경우 VenueHallIdNotExistsException 예외 처리
* VenueUpdateRequest DTO에 updateHallIds 와 VenueHallRequest를 VenueHallUpdateRequest로 통합
   =>만약 클라이언트에서 updateHallIds 속성 1개를 보냈는데 VenueHallUpdateRequest에서는 속성 2개를 보낼 시 서버에서 로직을 처리해 줘야 함
     따라서 통합하는 게 더 가독성이 좋아 보여 통합
* 수정, 삭제할 공연 홀 id가 DB에 존재하는지 로직 처리
    1. VenueHallService의 checkVenueHallIdsExists로 수정,삭제할 공연 홀 id들이 DB에 존재하는지 로직 처리
    2. VenueHallMapper의 getVenueHallCount로 업데이트할 공연 홀 id들이 DB에 몇개 존재하는지 반환 시켜줌
    3. 총 갯수가 아니라면 예외 처리
* 공연장 공연 홀 테스트 기능 추가

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/VenueHallService.java:
```diff
@@ -37,25 +40,60 @@ public void checkDuplicationVenueHallName(List<VenueHallRequest> venueHallReques
+    public void checkDuplicationVenueHallUpdateRequestNames(List<VenueHallUpdateRequest> venueHallUpdateRequests) {
+        Map<String, Integer> hallNameMap = new HashMap<>();
+
+        for(VenueHallUpdateRequest venueHallUpdateRequest : venueHallUpdateRequests) {
```
이건 stream 연산으로 대체가 가능해보입니다~

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/VenueHallService.java:
```diff
@@ -37,25 +40,60 @@ public void checkDuplicationVenueHallName(List<VenueHallRequest> venueHallReques
+    public List<Integer> getVenueHallUpdateRequestId(List<VenueHallUpdateRequest> venueHallUpdateRequests) {
+        List<Integer> hallIds = new ArrayList<>();
+
+        for(VenueHallUpdateRequest venueHallUpdateRequest : venueHallUpdateRequests) {
```
여기도 stream을 쓸 수 있을 것 같네요

tlsgkr7416 replied:
stream을 사용하여서 collection과 stream의 차이를 비교해 봤습니다.
collection은 메모리에 값을 저장하는 목표를 두고 있어서 collection에 저장하기 위해서는 미리 연산을 하여 저장을 합니다. 그리고 이때 연산 작업이 외부 반복문을 통해 이루어질 수 있습니다. 따라서 멀티 스레드를 적용한다면 동기화를 해줘야 합니다.
하지만 stream은 함수형 프로그래밍으로서 연산 작업 후 결과값을 출력하는 데 목표를 두고 있어 고정된 값(추가, 삭제가 불가능)만 출력이 됩니다. 그리고 내부 반복문을 사용하여 멀티 스레드가 적용되어도 동기화 작업을 하지 않고 스레드 세이프 하다는 장점이 있습니다.
마지막으로 collection처럼 for 문을 사용할 필요가 없기 때문에 가독성에도 좋다는 장점이 있습니다.

tlsgkr7416 committed:
[#50] 가독성 이유로 stream 적용

tlsgkr7416 committed:
[#41] 가독성 이유로 동일한 공연 홀 로직 stream 적용

tlsgkr7416 committed:
Merge branch 'feature/41/sh' into feature/50/sh

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/VenueHallService.java:
```diff
@@ -73,13 +75,10 @@ public void updateVenueHalls(int venueId, List<VenueHallUpdateRequest> venueHall
-        return hallIds;
+        return venueHallUpdateRequests.stream()
+                .map(venueHallUpdateRequest -> venueHallUpdateRequest.getId())
+                .collect(Collectors.toList());
```
👍

tlsgkr7416 replied:
approve해도 괜찮을까요?

tlsgkr7416 commented:
approve해도 괜찮을까요?

tlsgkr7416 committed:
Merge branch 'develop' into feature/50/sh

f-lab-dev reviewed (approved):
네 제가 실수로 approve를 comment로 잘못달았네요~
